### PR TITLE
[TMVA/SOFIE] Save weights in ROOT binary files

### PR DESCRIPTION
**Pull request by axmat**
## This Pull request:
This PR is based on #13066 by @Neel-Shah-29 

## Changes or fixes:
- A weight file isn't created by default when generating the code
- Options to save the weights in a text file or in a ROOT binary file

## Checklist:

- [x] tested changes locally

### CODE REVIEW

File: tmva/sofie/src/RModel.cxx [hunk 1]
```diff
@@ -459,54 +476,120 @@ namespace SOFIE{
...
-            fGC += "    }\n";
-            fGC += "   if (length != " + slength + ") {\n";
-            fGC += "      std::string err_msg = \"TMVA-SOFIE failed to read the correct tensor size; expected size is " +
-                   slength + " , read \" + std::to_string(length) ;\n";
-            fGC += "      throw std::runtime_error(err_msg);\n";
-            fGC += "    }\n";
-            fGC += "    for (int i =0; i < length; ++i) \n";
-            fGC += "       f >> " + tensor_name + "[i];\n";
+            fGC += "fTensor_" + i.first + " = *reinterpret_cast<std::vector<float>*>(rootFile->Get(";
+            fGC += "\"" + tensor_name + "\"));\n";
```
Comment (lmoneta) [inline tmva/sofie/src/RModel.cxx:528]: we should add here also int64_t and double since we have in writing

File: tmva/sofie/inc/TMVA/RModel.hxx [hunk 2]
```diff
@@ -52,7 +57,7 @@ private:
    const std::unordered_set<std::string> fAllowedStdLib = {"vector", "algorithm", "cmath"};
    std::unordered_set<std::string> fNeededStdLib = {"vector"};
    std::unordered_set<std::string> fCustomOpHeaders;
-   bool fUseWeightFile = true;
+   bool fUseWeightFile = false;
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/RModel.hxx:60]: I think the default here should be `true` and maybe use a .dat file as default

**Review by lmoneta — COMMENTED** [review]
Thank you for the PR!

File: tmva/sofie/inc/TMVA/RModel.hxx [hunk 3]
```diff
@@ -24,14 +24,20 @@ enum class Options {
    kDefault = 0x0,
    kNoSession = 0x1,
    kNoWeightFile = 0x2,
+   kTextWeightFile = 0x4,
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/RModel.hxx:27]: I don't think we need to add kTextWeightFile. This should be included in the kDefault option

**Review by lmoneta — COMMENTED** [review]
LGTM!
Just one comment for the enumeration defining the options

File: tmva/sofie/inc/TMVA/RModel.hxx (hunk 3 above)
Comment (axmat) [reply tmva/sofie/inc/TMVA/RModel.hxx]: Done, yes we don't need `kTextWeightFile` since its already included in `kDefault`.

**Review by lmoneta — APPROVED** [review]
LGTM!
Thank you Ahmat and Neel for this contribution

**Comment by lmoneta** [comment]
@phsft-bot build on ROOT-ubuntu2204/default, ROOT-ubuntu2004/default with flags -Dtmva-sofie=ON

**Pull request merged**